_posts/2018-12-06.Rust-1.31.md

```diff
@@ -0,0 +1,338 @@
+clear as to what is going on.
+
+Here's a brief summary:
+
+* `extern crate` is no longer needed in 99% of circumstances.
+* You can import macros with `use`, rather than a `#[macro_use]` attribute.
+* Absolute paths begin with a crate name, where the keyword `crate` refers to the current crate.
+* A `foo.rs` and `foo/` subdirectory may coexist; `mod.rs` is no longer needed when placing submodules in a subdirectory.
+
+These may seem like arbitrary new rules when put this way, but the mental
+model is now significantly simplified overall.
+
+There's a *lot* of details here, so please read [the edition
+guide](https://doc.rust-lang.org/nightly/edition-guide/rust-2018/module-system/path-clarity.html)
+for full details.
+
+### More lifetime elision rules
+
+Let's talk about a feature that's available in both editions: we've added
+some additional elision rules for `impl` blocks and function definitions.
+Code like this:
+
+```rust
+impl<'a> Reader for BufReader<'a> {
+    // methods go here
+}
+`
+
+can now be written like this:
+
+```rust
+impl Reader for BufReader<'_> {
+    // methods go here
+}
+```
+
+The `'_` lifetime still shows that `BufReader` takes a parameter, but we
+don't need to create a name for it anymore.
+
+Lifetimes are still required to be defined in structs. We're considering
+some options here in the future, but have no concrete plans yet.
+
+### `const fn`
+
+There's several ways to define a function in Rust: a regular function with
+`fn`, an unsafe function with `unsafe fn`, an external function with `extern fn`.
+This release adds a new way to qualify a function: `const fn`. It looks like
+this:
+
+```rust
+const fn foo(x: i32) -> i32 {
+    x + 1
+}
+```
+
+A `const fn` can be called like a regular function, but it can also be used
+in any constant context. When it is, it is evaluated at compile time, rather
+than at run time. As an example:
+
+```rust
+const SIX: i32 = foo(5);
+```
+
+This will execute `foo` at compile time, and set `SIX` to `6`.
+
+`const fn`s cannot do everything that normal `fn`s can do; they must
+have deterministic output. This is important for soundness reasons.
+Currently, `const fn`s can do a minimal subset of operations. Here's
+some examples of what you can do:
+
+* Arithmetic and comparison operators on integers
+* All boolean operators except for `&&` and `||`
+* Constructing arrays, structs, enums, and tuples
+* Calls to other `const fn`s
+* Index expressions on arrays and slices
+* Field accesses on structs and tuples
+* Reading from constants (but not statics, not even taking a reference to a static)
+* `&` and `*` of references
+* Casts, except for raw pointer to integer casts
+
+We'll be slowly growing the abilities of `const fn`, but we've decided that
+this is enough useful stuff to start shipping the feature itself.
+
+For full details, please see [the
+reference](https://doc.rust-lang.org/reference/items/functions.html#const-functions).
+
```

Comment:
```suggestion
The 2018 edition signals a new level of maturity for Rust's tools ecosystem. Cargo, Rustdoc, and Rustup have been crucial tools since 1.0; with the 2018 edition, there is a new generation of tools ready for all users: Clippy, Rustfmt, and IDE support.

--- Clippy (Manish) goes here ---

[Rustfmt](https://github.com/rust-lang/rustfmt) is a tool for formatting Rust code. Automatically formatting your code lets you save time and arguments by using the [official Rust style](https://github.com/rust-lang/rfcs/blob/master/style-guide/README.md). You can install with `rustup component add rustfmt` and use it with `cargo fmt`.

This release includes Rustfmt 1.0. From now on we guarantee backwards compatibility for Rustfmt: if you can format your code today, then the formatting will not change in the future (only with the default options). Backwards compatibility means that running Rustfmt on your CI is practical (use `cargo fmt --check`). Try that and 'format on save' in your editor to revolutionize your workflow.

IDE support is one of the most requested tooling features for Rust. There are now multiple, high quality options:

* [Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust)
* [IntelliJ](https://plugins.jetbrains.com/plugin/8182-rust)
* [Atom](https://github.com/rust-lang-nursery/atom-ide-rust)
* [Sublime Text 3](https://github.com/rust-lang/rust-enhanced)
* [Eclipse](https://www.eclipse.org/downloads/packages/release/photon/r/eclipse-ide-rust-developers-includes-incubating-components)

Work on IDE support is not finished, in particular code completion is not up to scratch in the RLS-based editors. However, if you mainly want support for types, documentation, and 'go to def', etc. then you should be happy.
```

Reply:
Stuff for clippy is in https://github.com/rust-lang/blog.rust-lang.org/pull/295#discussion_r239464428 (I can't add a suggestion on a suggestion).